tests/commons.py

```diff
@@ -15,18 +15,28 @@ class AsyncIterator:
     Async documents generator fake class, which records the args and kwargs it was called with.
     """
 
-    def __init__(self, items):
+    def __init__(self, items, reusable=False):
+        """
+        AsyncIterator is a test-only abstraction to mock async iterables.
+        By default it's usable only once: once iterated over, he iterator will not
+        iterate again any more.
+        If reusable is True, then iterator can be re-used, but only if it's used by a single coroutine.
+        If AsyncIterator is used in several coroutines, it'll not work correctly at all
+        """
         self.items = items
         self.call_args = []
         self.call_kwargs = []
         self.i = 0
         self.call_count = 0
+        self.reusable = reusable
 
     def __aiter__(self):
         return self
 
     async def __anext__(self):
         if self.i >= len(self.items):
+            if self.reusable:
+                self.i = 0
```

Comment:
I can see it's only for tests so not a big deal but does this open a risk for infinite looping anywhere?

Reply:
It should not be, no. Here there's no return statement, so next line that raises an error will be triggered and iterations will stop. Next time this iterator is used it's going to be iterating over same original collection, before it would immediately stop iteration.